File: python/BOJ/BOJ13913.py
import sys
from queue import Queue

def bfs(st,en):
    que = Queue()
    que.put((st,0))
    visited[st] = True

    while not(que.empty()):
        now = que.get()

        if now[0] == en:
            return now[1]

        # nxt pos
        nxt = now[0] + 1
        if nxt <= 100000 and not(visited[nxt]):
            visited[nxt] = True
            before[nxt] = now[0]
            que.put((nxt,now[1] + 1))

        nxt = now[0] - 1
        if nxt >= 0 and not(visited[nxt]):
            visited[nxt] = True
            before[nxt] = now[0]
            que.put((nxt,now[1] + 1))

        nxt = 2 * now[0]
        if nxt <= 100000 and not(visited[nxt]):
            visited[nxt] = True
            before[nxt] = now[0]
            que.put((nxt,now[1] + 1))

subin, brother = map(int,sys.stdin.readline().split())
visited = [False for loop in range(100001)]
before = [0 for loop in range(100001)]

answer = bfs(subin,brother)
sys.stdout.write(str(answer) + '\n')

posList = []
pos = brother
while pos != subin:
    posList.append(pos)
    pos = before[pos]

posList.append(subin)
posList.reverse()

for pos in posList:
    sys.stdout.write(str(pos) + " ")

File: java/BOJ/BOJ13913.java
package boj;

import java.io.*;
import java.util.*;

class Main {
	
	static class Pair {
		int first, second;
		
		Pair(int f, int s) {
			first = f;
			second = s;
		}
	}
	
	static final int POSITION_MAX = 100001;
	static int[] parent = new int[POSITION_MAX];
	
	public static boolean isSafe(int x) {
		return 0 <= x && x < POSITION_MAX;
	}
	
	public static int[] getNxtPos(int pos) {
		int[] result = {pos-1,pos+1,2*pos};
		return result;
	}
	
	public static int bfs(int st, int en) {
		
		boolean[] visited = new boolean[POSITION_MAX];
		Queue<Pair> q = new LinkedList<>();
		q.add(new Pair(st,0));
		
		while (!q.isEmpty()) {
			
			Pair now = q.poll();
			
			if (now.first == en) {
				return now.second;
			}
			
			int[] nxtPos = getNxtPos(now.first);
			
			for (int pos : nxtPos) {
				if (isSafe(pos) && !visited[pos]) {
					visited[pos] = true;
					parent[pos] = now.first;
					q.add(new Pair(pos,now.second + 1));
				}
			}
		}
		
		return -1;
	}
	
	//BOJ13913
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int n,k;
		String[] pos = br.readLine().split(" ");
		n = Integer.parseInt(pos[0]);
		k = Integer.parseInt(pos[1]);
		
		System.out.println(bfs(n,k));
		StringBuilder sb = new StringBuilder();
		
		int now = k;
		List<Integer> result = new ArrayList<>();
		while (now != n) {
			result.add(now);
			now = parent[now];
		}
		result.add(n);
		Collections.reverse(result);
		
		for (int position : result) {
			sb.append(position);
			sb.append(" ");
		}
		
		System.out.println(sb.toString());
	}
}
